In [1]:
cd /home/lizichen/Downloads

/home/lizichen/Downloads


In [2]:
import h5py
import numpy as np
import collections
import datetime
import csv
import pickle

/home/lizichen/anaconda3/envs/modal/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [ ]:
f = h5py.File('features.h5', 'r')

In [ ]:
cluster_recording = list(f.values())[0]
cluster_recording.shape

## Raw Data Peeking: The first ten 3-seconds audio clips:

In [ ]:
cluster_recording[1][0]

In [ ]:
cluster_recording[1][1].decode("utf-8").split('_')

In [ ]:
embedding = cluster_recording[1][2]
embedding

In [ ]:
embedding.shape

## 1. Convert the plain data format to dictionary for fast key look-up

In [ ]:
alldata = collections.defaultdict(dict)

numberOfDataPoints = cluster_recording.shape[0]

counter = 0
print('Start: '+str(datetime.datetime.now()))

for datapoint in range(numberOfDataPoints):
    sensor_and_timestamp = cluster_recording[datapoint][1].decode("utf-8")
    if len(sensor_and_timestamp.split('_')) == 2:
        sensor_id = cluster_recording[datapoint][1].decode("utf-8").split('_')[0]
        timestamp = cluster_recording[datapoint][1].decode("utf-8").split('_')[1]
        embedding = cluster_recording[datapoint][2]
        alldata[sensor_id][timestamp] = embedding
    elif len(sensor_and_timestamp.split('-')) == 2:
        sensor_id = cluster_recording[datapoint][1].decode("utf-8").split('-')[0]
        timestamp = cluster_recording[datapoint][1].decode("utf-8").split('-')[1]
        embedding = cluster_recording[datapoint][2]
        alldata[sensor_id][timestamp] = embedding
    else:
        print('Warning: Not splittable!:' + sensor_and_timestamp)
    
    counter += 1
        
    if counter % 150000 == 0:
        print(str(counter) + ' ' + str(datetime.datetime.now()))

### 1.1 Save the re-formatted feature data; now in the format of DefaultDictionary, into pickle file for efficient reading

In [ ]:
with open("features.pickle", 'wb') as pfile:
    pickle.dump(alldata, pfile, protocol=pickle.HIGHEST_PROTOCOL)

### 1.2 Load saved dictionary pickle! (All features data)

In [3]:
# alldata_dict = collections.defaultdict(dict)
with (open("features.pickle", "rb")) as openfile:
    while True:
        try:
            alldata_dict= pickle.load(openfile)
        except EOFError:
            break

### 1.3 Example: Now we have all the sensor IDs and timestamp effectively accessible in the dictionary

In [4]:
alldata_dict.keys()

dict_keys(['74da385c6855', 'b827ebb40450', 'b827eb4cc22e', 'b827eb84deb5', 'b827eb44506f', 'b827eb0d8af7', 'b827eb7b2c3e', 'b827eb252949', 'b827eb977bfb', 'b827ebe1fe4b', 'b827eb9bed23', 'b827eb5d1714', 'b827eb539980', '0013ef2008e6', 'b827eb2a1bce', 'b827eb3bda47', 'b827eb86d458', 'b827ebad073b', 'b827eb29eb77', 'b827eb820cfe', '0013ef6b0a0f', 'b827eb2c65db', 'b827ebfd616c', 'b827eb8e2420', 'b827eb905497', '74da385c684f', '0013ef700444', 'b827eb122f0f', 'b827eb32f75c', 'b827eb429cd4', '74da385c683d', 'b827ebba613d', 'b827eb3e842e', '0013ef801412', 'b827eb0fedda', 'b827eb5895e9', 'b827ebc7f772', 'b827ebf9d204', 'b827eb73e772', '74da385c687d', 'b827eb4e7821', 'b827eb9d0e7f', '74da383b5ca4', 'b827eb1685c7', 'b827eb815321', 'b827ebefb215', 'b827eb43d8f4', 'b827ebdd5c38', 'b827eb8e1f0b', 'b827eb8e32ad', 'b827eb132382', 'b827eb42bd4a', 'b827ebc6dcc6', 'b827eb329ab8', 'b827ebf31214'])

In [28]:
alldata_dict['74da385c6855'].keys()

dict_keys(['1484629333.39', '1486275276.49', '1486270823.64', '1484456462.61', '1484506256.94', '1484456868.87', '1485062169.07', '1485061248.92', '1485107944.57', '1485105300.84', '1484370845.88', '1484370025.38', '1484377939.68', '1484377520.33', '1484802567.76', '1484812312.72', '1484802064.4', '1484811876.07', '1484715649.85', '1485320440.89', '1485397215.11', '1486457713.33', '1486510104.53', '1486509851.17', '1485493226.56', '1485982646.26', '1484888890.45', '1484888476.2', '1484974865.71', '1485255027.89', '1485241169.45', '1485249454.09'])

## 2. Read the positive_samples.pickle file - all positive sensor_id and timestamp
positive_samples.pickle file from Yu Wang, April 6, 2018, shared on Slack group.

In [7]:
# alldata_dict = collections.defaultdict(dict)
with (open("positive_samples.pickle", "rb")) as openfile:
    while True:
        try:
            positive_samples= pickle.load(openfile)
        except EOFError:
            break

In [8]:
positive_samples

{'b827eb429cd4': ['1488836836.42', '1488810598.15'],
 'b827eb5895e9': ['1490690898.85', '1490686108.53'],
 'b827eb0d8af7': ['1485198017.91', '1485163671.07'],
 'b827eb122f0f': ['1498892374.76', '1498261062.11'],
 'b827eb86d458': ['1487000875.51', '1487028598.04'],
 'b827eb0fedda': ['1497928649.89', '1480740512.58'],
 'b827eb8e2420': ['1500149760.47', '1500173615.97'],
 'b827eb1685c7': ['1499724045.58', '1486212620.04'],
 'b827eb815321': ['1482308965.46', '1482399908.63'],
 'b827eb9bed23': ['1491297344.73', '1491323380.55'],
 'b827eb44506f': ['1486149582.98', '1499473078.4'],
 'b827eb4e7821': ['1483014176.71', '1483015339.86'],
 'b827eb2a1bce': ['1487029980.37', '1487018417.81'],
 'b827ebad073b': ['1480303515.84', '1480297488.98'],
 'b827eb42bd4a': ['1495845294.38', '1495820507.31']}

### 2.1 Select the positive embeddings according to the csv file data  - Set up positive training dataset

In [15]:
totalPositiveEmbedding = np.empty([1,128], dtype=int)
for sensorid in positive_samples.keys():
    for timestamp in positive_samples[sensorid]:
        if sensorid in alldata_dict.keys():
            if timestamp in alldata_dict[sensorid].keys():
                embedding = alldata_dict[sensorid][timestamp]
                totalPositiveEmbedding = np.vstack((totalPositiveEmbedding, embedding))
        
totalPositiveEmbedding =np.delete(totalPositiveEmbedding, 0, 0) # remove the first dummy 128 embedding

In [16]:
totalPositiveEmbedding.shape

(170, 128)

In [17]:
# add the positive label, 1
positive_xy = np.insert(totalPositiveEmbedding, 128, 1, axis=1)

In [19]:
positive_xy.shape

(170, 129)

### 2.2 Save the positive training embedding data to pickle file:

In [21]:
with open("positive_xy.pickle", 'wb') as pfile:
    pickle.dump(positive_xy, pfile, protocol=pickle.HIGHEST_PROTOCOL)

### 2.3 Load the pickle! (positive training numpy object 170 $\times$ (128+1), with 1 mark as Is-Noise):

In [22]:
with (open("positive_xy.pickle", "rb")) as openfile:
    while True:
        try:
            positive_xy= pickle.load(openfile)
        except EOFError:
            break

## 3. Read the negative csv file

In [33]:
negativeCsvData = {}
with open('Sensor_data_negative.csv') as csvfile:
    Sensor_data_positive = csv.reader(csvfile, skipinitialspace=True, delimiter=' ')
    csvrow = -1
    for row in Sensor_data_positive:
        if csvrow != -1: # skip the first row
            sid = row[0].split(',')[0]
            ts  = row[0].split(',')[1]
            if sid in negativeCsvData.keys():
                negativeCsvData[sid].append(ts)
            else:
                negativeCsvData[sid] = []
                negativeCsvData[sid].append(ts)
        csvrow += 1   

In [50]:
print(type(negativeCsvData))
print(len(negativeCsvData))
print(negativeCsvData['b827eb4e7821'])

<class 'dict'>
15
['1491211362.82_5', '1491800203.12_8', '1482424939.17_1', '1498868810.69_4', '1488993967.2_1', '1492066899.36_6', '1484297539.06_6', '1490406200.26_1', '1494789341.92_4', '1490926970.82_4', '1488422336.29_2', '1488377458.61_6', '1486479347.85_4', '1483636861.26_6', '1484122883.92_3', '1487404654.04_8', '1492186534.34_2', '1493876183.72_2', '1483134229.53_1', '1491421469.99_7']


### 3.1 Pull negative embeddings according to the negative sensor and timestamp data (negativeCsvData)

In [43]:
totalNegativeEmbedding = np.empty([1,128], dtype=int)
for sensorid in negativeCsvData.keys():
    for timestamp in negativeCsvData[sensorid]:
        if sensorid in alldata_dict.keys():
            all_timestamp_full = negativeCsvData[sensorid]
            for timestamp_full in all_timestamp_full:
                timestamp = timestamp_full.split('_')[0]
                embedding_index = timestamp_full.split('_')[1]
                if timestamp in alldata_dict[sensorid]:
                    one_negative_embedding = alldata_dict[sensorid][timestamp][int(embedding_index)]
                    totalNegativeEmbedding = np.vstack((totalNegativeEmbedding, one_negative_embedding))
            
totalNegativeEmbedding =np.delete(totalNegativeEmbedding, 0, 0) # remove the first dummy 128 embedding

In [55]:
print(totalNegativeEmbedding.shape)
print(totalNegativeEmbedding)

(3980, 128)
[[146  14 129 ...   0  62 255]
 [154  30 171 ...  79 158 255]
 [153  25 150 ... 114 150 255]
 ...
 [164  33 165 ... 211 244 255]
 [164  37 180 ... 174 166 255]
 [153  28 167 ...  76 108 255]]


### 3.2 Add negative labels (0) to the negative embeddings.

In [51]:
# add the negative label, 0
negative_xy = np.insert(totalNegativeEmbedding, 128, 1, axis=1)

In [52]:
negative_xy.shape

(3980, 129)

### 3.3 Save the negative embedding and labels to a pickle file.

In [56]:
with open("negative_xy.pickle", 'wb') as pfile:
    pickle.dump(negative_xy, pfile, protocol=pickle.HIGHEST_PROTOCOL)

### 3.4 Load the negative_xy pickle

In [57]:
with (open("negative_xy.pickle", "rb")) as openfile:
    while True:
        try:
            negative_xy= pickle.load(openfile)
        except EOFError:
            break

## 4. Summary:
Now we have the following datasets that have been cleaned and formatted to pickles for better usage:  
1. ``features.pickle``
2. ``positive_xy.pickle``
3. ``negative_xy.pickle``

The original raw data files are:
1. ``features.h5``
2. ``positive_samples.pickle``
3. ``Sensor_data_negative.csv``